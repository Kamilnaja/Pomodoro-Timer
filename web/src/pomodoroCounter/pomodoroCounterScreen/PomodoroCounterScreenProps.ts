import { AuthState } from '../../auth/store/interfaces/authState';
import { PomodoroCounterState } from '../store/interfaces/PomodoroCounterState';

export interface PomodoroCounterScreenProps {
  auth: AuthState;
  pomodoroCounter: PomodoroCounterState;
  handleSavePomodoro: () => void;
  handleSetModePomodoro: () => void;
  handleSetModeBreak: (time: number) => void;
}
