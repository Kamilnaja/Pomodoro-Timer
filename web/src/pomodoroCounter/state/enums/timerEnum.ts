export enum CounterState {
  POMODORO_RUNNING = 'POMODORO_RUNNING',
  BREAK_RUNNING = 'BREAK_RUNNING',
  POMODORO_END = 'POMODORO_END',
  BREAK_END = 'BREAK_END',
  POMODORO_PAUSE = 'POMODORO_PAUSE',
  BREAK_PAUSE = 'BREAK_PAUSE',
}
