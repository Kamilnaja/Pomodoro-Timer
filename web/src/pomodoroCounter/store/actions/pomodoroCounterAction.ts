import { Action } from 'redux';
import { ActionWithPayload } from '../../../store/interfaces/actions/actionInterface';

export enum PomodoroCounterActions {
  PAUSE = 'PAUSE',
  RUN = 'RUN',
  END = 'END',
  UPDATE_TIME = 'UPDATE_TIME',
  SET_MODE_POMODORO = 'SET_MODE_POMODORO',
  SET_MODE_BREAK = 'SET_MODE_BREAK',
}

export const pause = (): Action<PomodoroCounterActions> => ({
  type: PomodoroCounterActions.PAUSE,
});

export const run = (): Action<PomodoroCounterActions> => ({
  type: PomodoroCounterActions.RUN,
});

export const end = (): Action<PomodoroCounterActions> => ({
  type: PomodoroCounterActions.END,
});

export const updateCounter = (payload: number): ActionWithPayload<PomodoroCounterActions, number> => ({
  type: PomodoroCounterActions.UPDATE_TIME,
  payload,
});

export const setModePomodoro = (): Action<PomodoroCounterActions> => ({
  type: PomodoroCounterActions.SET_MODE_POMODORO,
});

export const setModeBreak = (): Action<PomodoroCounterActions> => ({
  type: PomodoroCounterActions.SET_MODE_BREAK,
});

// thunk
export const handleSetModePomodoro = () => (dispatch: (action: Action<PomodoroCounterActions>) => void) => {
  dispatch(end());
  dispatch(setModePomodoro());
};

export const handleSetModeBreak = () => (dispatch: (action: Action<PomodoroCounterActions>) => void) => {
  dispatch(end());
  dispatch(setModeBreak());
};
