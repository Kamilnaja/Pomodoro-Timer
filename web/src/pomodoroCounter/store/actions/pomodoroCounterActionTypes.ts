export const PAUSE = 'PAUSE';
export const RUN = 'RUN';
export const END = 'END';
export const UPDATE_TIME = 'UPDATE_TIME';
export const SET_MODE_POMODORO = 'SET_MODE_POMODORO';
export const SET_MODE_BREAK = 'SET_MODE_BREAK';

export interface PauseAction {
  type: typeof PAUSE;
}

export interface RunAction {
  type: typeof RUN;
}

export interface EndAction {
  type: typeof END;
}

export interface UpdateTimeAction {
  type: typeof UPDATE_TIME;
  payload: number;
}

export interface SetModePomodoroAction {
  type: typeof SET_MODE_POMODORO;
}

export interface SetModeBreakAction {
  type: typeof SET_MODE_BREAK;
  payload: number;
}

export type PomodoroCounterActionsTypes =
  | PauseAction
  | RunAction
  | EndAction
  | UpdateTimeAction
  | SetModePomodoroAction
  | SetModeBreakAction;
