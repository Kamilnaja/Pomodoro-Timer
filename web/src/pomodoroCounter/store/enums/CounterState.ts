export enum CounterState {
  RUNNING = 'RUNNING',
  END = 'END',
  PAUSE = 'PAUSE',
  CURRENT_ACTION = 'CURRENT_ACTION',
}
