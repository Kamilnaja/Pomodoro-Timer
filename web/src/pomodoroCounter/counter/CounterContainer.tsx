import React from 'react';
import { Button } from 'react-bootstrap';
import { connect } from 'react-redux';
import { msToTime } from '../../shared/scripts/utils';
import { initialConfig } from '../../shared/settings/initialConfig';
import { TabTitle } from '../../shared/title/TabTitle';
import {
  isAnyTimerRunning,
  playClickSound,
  isPomodoroMode,
} from '../pomodoroCounterScreen/PomodoroCounterScreenHelpers';
import { end, pause, run, updateCounter } from '../store/actions/pomodoroCounterAction';
import { PomodoroCounterState } from '../store/interfaces/PomodoroCounterState';
import './counterContainer.scss';
import { CounterComponentProps } from './CounterContainerProps';
import { worker } from './Worker';

class CounterContainer extends React.Component<CounterComponentProps> {
  private tabTitle = new TabTitle();
  private worker: Worker = worker;

  componentDidMount() {
    this.initializeWorker();
  }

  handlePauseCounter = () => {
    this.props.pause();
    this.setWorkerTime();
  };

  handleStartCounter = () => {
    if (!isAnyTimerRunning(this.props.counter.counterState)) {
      this.tabTitle.stopBlinking();
      playClickSound();

      this.props.run();

      this.worker.postMessage({
        type: 'START',
        payload: this.props.counter.counterTime,
      });
    }
  };

  handleStartNewPomodoro = () => {
    this.setWorkerTime(initialConfig.pomodoroTime);
    this.props.end();
  };

  handleStartNewBreak = (time: number) => {
    this.setWorkerTime(time);
    this.props.end();
  };

  handlePomodoroEnd = () => {
    this.props.handleSavePomodoro();
  };

  private setWorkerTime = (time: number = 0) => {
    this.worker.postMessage({
      type: 'SET_TIME',
      payload: time,
    });
  };

  private initializeWorker() {
    if (typeof Worker !== 'undefined')
      this.worker.onmessage = e => {
        const { data } = e;
        if (data.isRunning) {
          const { time } = data;
          this.tabTitle.setTitle = `${msToTime(time)}`;
          this.props.updateCounter(time);
        } else {
          this.tabTitle.startBlinking();

          if (isPomodoroMode(this.props.counter.currentTimer)) {
            this.props.handleSavePomodoro();
            this.setWorkerTime(initialConfig.shortBreakTime);
          } else {
            console.log('break end!');
          }
        }
      };
  }

  render() {
    return (
      <>
        <p className="time text-light d-flex align-items-center">{msToTime(this.props.counter.counterTime)}</p>
        {isAnyTimerRunning(this.props.counter.counterState) ? (
          <Button variant="outline-primary" onClick={this.handlePauseCounter}>
            Pause timer
          </Button>
        ) : (
          <Button variant="success" onClick={this.handleStartCounter}>
            Start timer
          </Button>
        )}
      </>
    );
  }
}

const mapStateToProps = (state: { pomodoroCounter: PomodoroCounterState }) => {
  const { pomodoroCounter } = state;
  return { counter: pomodoroCounter };
};

const mapDispatchToProps = {
  updateCounter,
  pause,
  run,
  end,
};

export default connect(mapStateToProps, mapDispatchToProps)(CounterContainer);
