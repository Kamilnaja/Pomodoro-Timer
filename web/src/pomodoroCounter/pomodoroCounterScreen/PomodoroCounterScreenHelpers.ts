import { CounterState } from '../store/enums/CounterState';
import { CurrentTimer } from '../store/interfaces/PomodoroCounterState';

export const playClickSound = () => {
  const audio = new Audio('sounds/zapsplat_multimedia_game_sound_childrens_ping_high_pitched_soft_007_60676.mp3');
  audio.play();
};

export const playEndSound = () => {
  const audio = new Audio('sounds/zapsplat_multimedia_game_sound_positive_award_bonus_bright_warm_synth_001_60698.mp3');
  audio.play();
};

export const isAnyTimerRunning = (counterState: CounterState): boolean => counterState === CounterState.RUNNING;

export const isPomodoroMode = (currentTimer: CurrentTimer): boolean => currentTimer === CurrentTimer.POMODORO;
