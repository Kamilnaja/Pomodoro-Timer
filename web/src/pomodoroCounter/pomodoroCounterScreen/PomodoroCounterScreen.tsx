import React from 'react';
import { Jumbotron } from 'react-bootstrap';
import { connect } from 'react-redux';
import 'shared/settings/initialConfig';
import { AuthState } from '../../auth/store/interfaces/authState';
import { handleSavePomodoro } from '../../stats/store/actions/statsActions';
import CounterContainer from '../counter/CounterContainer';
import { InfoComponent } from '../info/InfoComponent';
import { ModeButtonsComponent } from '../modeButtons/ModeButtonsComponent';
import { PomodoroCounterState } from '../store/interfaces/PomodoroCounterState';
import { timerState } from '../store/state/timerState';
import { PomodoroCounterScreenProps } from './PomodoroCounterScreenProps';
import { handleSetModeBreak, handleSetModePomodoro } from '../store/actions/pomodoroCounterAction';
import { initialConfig } from 'shared/settings/initialConfig';
import { worker } from '../counter/Worker';
import { TabTitle } from '../../shared/title/TabTitle';

class PomodoroCounterScreen extends React.Component<PomodoroCounterScreenProps> {
  private worker: Worker = worker;
  private tabTitle: TabTitle = new TabTitle();

  constructor(props: PomodoroCounterScreenProps) {
    super(props);
    this.state = timerState;
  }

  handleSetModePomodoro = () => {
    this.props.handleSetModePomodoro();
    worker.postMessage({
      type: 'SET_TIME',
      payload: initialConfig.pomodoroTime,
    });
  };

  handleSetModeLongBreak = () => {
    this.props.handleSetModeBreak(initialConfig.longBreakTime);
    worker.postMessage({
      type: 'SET_TIME',
      payload: initialConfig.longBreakTime,
    });
  };

  handleSetModeShortBreak = () => {
    this.props.handleSetModeBreak(initialConfig.shortBreakTime);
    worker.postMessage({
      type: 'SET_TIME',
      payload: initialConfig.shortBreakTime,
    });
  };

  render = () => (
    <Jumbotron className="align-items-center d-flex flex-column bg-dark">
      <ModeButtonsComponent
        setModePomodoro={this.handleSetModePomodoro}
        setModeLongBreak={this.handleSetModeLongBreak}
        setModeShortBreak={this.handleSetModeShortBreak}
      />
      <CounterContainer handleSavePomodoro={this.props.handleSavePomodoro}></CounterContainer>
      <InfoComponent currentState={this.props.pomodoroCounter.counterState} auth={this.props.auth} />
    </Jumbotron>
  );
}
const mapDispatchToProps = {
  handleSavePomodoro,
  handleSetModePomodoro,
  handleSetModeBreak,
};

const mapStateToProps = (state: { auth: AuthState; pomodoroCounter: PomodoroCounterState }) => {
  const { auth, pomodoroCounter } = state;
  return { auth, pomodoroCounter };
};

export default connect(mapStateToProps, mapDispatchToProps)(PomodoroCounterScreen);
