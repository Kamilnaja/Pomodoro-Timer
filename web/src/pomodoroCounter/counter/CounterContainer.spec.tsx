import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { Store } from 'redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { pause, run } from '../store/actions/pomodoroCounterAction';
import { PomodoroCounterActionsTypes } from '../store/actions/pomodoroCounterActionTypes';
import { CounterState } from '../store/enums/CounterState';
import { createPomodoroCounterState } from '../testing/pomodoroCounter.test.data';
import CounterContainer from './CounterContainer';
import * as Helpers from '../pomodoroCounterScreen/PomodoroCounterScreenHelpers';

const mockStore = configureStore([thunk]);

describe('CounterContainer', () => {
  let store: Store<any, PomodoroCounterActionsTypes>;
  let component: renderer.ReactTestRenderer;
  let playClickSoundSpy: any;

  const handleSavePomodoro = jest.fn();

  beforeEach(() => {
    store = mockStore({
      pomodoroCounter: createPomodoroCounterState(),
    });

    store.dispatch = jest.fn();

    component = renderer.create(
      <Provider store={store}>
        <CounterContainer handleSavePomodoro={handleSavePomodoro} />
      </Provider>,
    );
  });

  it('Should match snapshot', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('Should pause counter when counter is running', () => {
    const storeSpy = jest.spyOn(store, 'dispatch');

    renderer.act(() => {
      component.root.findByProps({ className: 'btn--pause' }).props.onClick();
    });

    expect(storeSpy).toHaveBeenCalledWith(pause());
  });

  xit('Should start counter', () => {
    const storeSpy = jest.spyOn(store, 'dispatch');
    jest.spyOn(Helpers, 'playClickSound');

    store = mockStore({
      pomodoroCounter: createPomodoroCounterState(CounterState.END),
    });

    store.dispatch = jest.fn();

    component = renderer.create(
      <Provider store={store}>
        <CounterContainer handleSavePomodoro={handleSavePomodoro} />
      </Provider>,
    );

    renderer.act(() => {
      component.root.findByProps({ className: 'btn--start' }).props.onClick();
    });

    expect(storeSpy).toHaveBeenCalledWith(run());
    expect(playClickSoundSpy).toHaveBeenCalledTimes(1);
  });
});
