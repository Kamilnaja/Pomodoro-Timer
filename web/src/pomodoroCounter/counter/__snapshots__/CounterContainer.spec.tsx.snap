// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CounterContainer Should match snapshot 1`] = `
Array [
  <p
    className="time text-light d-flex align-items-center"
  >
    41:40
  </p>,
  <button
    className="btn--pause btn btn-outline-primary"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Pause timer
  </button>,
]
`;
